apps/remote_control/lib/lexical/remote_control/search/fuzzy.ex

```diff
@@ -0,0 +1,251 @@
+defmodule Lexical.RemoteControl.Search.Fuzzy do
+  @moduledoc """
+  A backend for fuzzy matching
+
+  This is a storage module that allows you to map keys to multiple values in two ways.
+
+  Values are grouped by their `subject`, which is a string that enables fuzzy matching. They can also be grouped
+  by a `grouping key` that allows for grouping by an arbitrary value.
+
+  For both cases, multiple values can exist under the same key, and when searching, all values under the key are
+  returned.
+  """
+
+  alias Lexical.RemoteControl.Search.Fuzzy.Scorer
+  alias Lexical.RemoteControl.Search.Indexer.Entry
+
+  defstruct subject_to_values: %{},
+            grouping_key_to_values: %{},
+            preprocessed_subjects: %{},
+            mapper: nil,
+            subject_converter: nil
+
+  @type subject :: String.t()
+  @type extracted_subject :: term()
+  @type grouping_key :: term()
+  @type value :: term()
+  @type extracted_subject_grouping_key_value :: {extracted_subject(), grouping_key(), value()}
+  @type mapper :: (term -> extracted_subject_grouping_key_value)
+  @type subject_converter :: (extracted_subject() -> subject())
+
+  @opaque t :: %__MODULE__{
+            subject_to_values: %{subject() => [value()]},
+            grouping_key_to_values: %{Path.t() => [value()]},
+            preprocessed_subjects: %{subject() => tuple()},
+            mapper: mapper(),
+            subject_converter: subject_converter()
+          }
+
+  @spec from_entries([Entry.t()]) :: t
+  def from_entries(entries) do
+    mapper = fn %Entry{} = entry ->
+      {entry.subject, entry.path, entry.ref}
+    end
+
+    new(entries, mapper, &stringify/1)
+  end
+
+  @doc """
+  Creates a new fuzzy matcher.
+
+  Items in the enumerable first argument will have the mapper function applied to them.
+  For each tuple returned, the first element will then have the subject converter applied.
+  This will produce a subject, which is what the `match/2` function uses for fuzzy matching.
+  """
+  @spec new([term()], mapper(), subject_converter()) :: t
+  def new(items, mapper, subject_converter) do
+    subject_grouping_key_values = Enum.map(items, mapper)
+
+    extract_and_fix_subject = fn {subject, _, _} -> subject_converter.(subject) end
+
+    subject_to_values =
+      Enum.group_by(subject_grouping_key_values, extract_and_fix_subject, &elem(&1, 2))
+
+    grouping_key_to_values =
+      Enum.group_by(subject_grouping_key_values, &elem(&1, 1), &elem(&1, 2))
+
+    preprocessed_subjects =
+      subject_to_values
+      |> Map.keys()
+      |> Map.new(fn subject -> {subject, Scorer.preprocess(subject)} end)
+
+    %__MODULE__{
+      subject_to_values: subject_to_values,
+      grouping_key_to_values: grouping_key_to_values,
+      preprocessed_subjects: preprocessed_subjects,
+      mapper: mapper,
+      subject_converter: subject_converter
+    }
+  end
```

Comment:
this is for performance; Calling Map.update tens of thousands of times is vastly, vastly slower than building a list of tuples and creating new maps.

Reply:
Makes sense!

Would it be worth defining add in terms of new, then? Add could create a new fuzzy struct and then combine them. My main concern is that the "prep" logic is non-trivial enough that it would be nice not to have to essentially duplicate it. 

Defining add in terms of new would probably also be a perf hit if we're adding a lot of single items, but it may improve perf when adding bulk items.

Reply:
> Defining add in terms of new would probably also be a perf hit

Yes, exactly. The only way it'd be faster is if you were adding more items than already exist in the data structure.